packages/firestore/src/local/indexeddb_document_overlay_cache.ts

```diff
@@ -0,0 +1,205 @@
+    collectionPaths.forEach(collectionPath => {
+      const range = IDBKeyRange.bound(
+        [this.userId, collectionPath, batchId],
+        [this.userId, collectionPath, batchId + 1],
+        /*lowerOpen=*/ false,
+        /*upperOpen=*/ true
+      );
+      promises.push(
+        documentOverlayStore(transaction).deleteAll(
+          DbDocumentOverlay.collectionPathOverlayIndex,
+          range
+        )
+      );
+    });
+    return PersistencePromise.waitFor(promises);
+  }
+
+  getOverlaysForCollection(
+    transaction: PersistenceTransaction,
+    collection: ResourcePath,
+    sinceBatchId: number
+  ): PersistencePromise<Map<DocumentKey, Overlay>> {
+    const result = new Map<DocumentKey, Overlay>();
+    const collectionPath = encodeResourcePath(collection);
+    // We want batch IDs larger than `sinceBatchId`, and so the lower bound
+    // is not inclusive.
+    const range = IDBKeyRange.bound(
+      [this.userId, collectionPath, sinceBatchId],
+      [this.userId, collectionPath, Number.POSITIVE_INFINITY],
+      /*lowerOpen=*/ true
+    );
+    return documentOverlayStore(transaction)
+      .loadAll(DbDocumentOverlay.collectionPathOverlayIndex, range)
+      .next(dbOverlays => {
+        for (const dbOverlay of dbOverlays) {
+          const overlay = fromDbDocumentOverlay(this.serializer, dbOverlay);
+          result.set(overlay.getKey(), overlay);
+        }
+        return result;
+      });
+  }
+
+  getOverlaysForCollectionGroup(
+    transaction: PersistenceTransaction,
+    collectionGroup: string,
+    sinceBatchId: number,
+    count: number
+  ): PersistencePromise<Map<DocumentKey, Overlay>> {
+    const result = new Map<DocumentKey, Overlay>();
+    let currentBatchId: number | undefined = undefined;
+    let currentCount = 0;
+    // We want batch IDs larger than `sinceBatchId`, and so the lower bound
+    // is not inclusive.
+    const range = IDBKeyRange.bound(
+      [this.userId, collectionGroup, sinceBatchId],
+      [this.userId, collectionGroup, Number.POSITIVE_INFINITY],
+      /*lowerOpen=*/ true
+    );
+    return documentOverlayStore(transaction)
+      .iterate(
+        {
+          index: DbDocumentOverlay.collectionGroupOverlayIndex,
+          range
+        },
+        (_, dbOverlay, control) => {
+          // We do not want to return partial batch overlays, even if the size
+          // of the result set exceeds the given `count` argument. Therefore, we
+          // continue to aggregate the results even after `currentCount` exceeds
+          // `count` if there are more overlays from the `currentBatchId`.
+          const overlay = fromDbDocumentOverlay(this.serializer, dbOverlay);
+          if (
+            currentCount < count ||
+            overlay.largestBatchId === currentBatchId
+          ) {
+            result.set(overlay.getKey(), overlay);
+            currentBatchId = overlay.largestBatchId;
+            ++currentCount;
```

Comment:
We can just use `result.size()` instead of current count.

Reply:
Done